Update: Document custom annotations for decK APIOps (#7098)
Kong/docs.konghq.com: app/_src/deck/guides/apiops.md

to bring full and partial {{site.base_gateway}} configurations together to create APIOps for your 
particular needs.

## Extending OpenAPI Specifications

> Several custom annotations can be used within the OpenAPI specification allowing you to declare {{site.base_gateway}} capabilities directly in the specification document. All custom annotations related to {{site.base_gateway}} configuration are prepended with ```x-kong-``` label.

Reviewer: ```suggestion
You can use several custom annotations within the OpenAPI specification, allowing you to declare {{site.base_gateway}} capabilities directly in the specification document. All custom annotations related to {{site.base_gateway}} configuration are prepended with the `x-kong-` label.
```

> - ```x-kong-tags```: specify the tags to use for each Kong entity generated. The tags can be overridden when doing the conversion. This can only be specified on document level.

Reviewer: Let's move these annotations into a table. Something like this: https://docs.konghq.com/kubernetes-ingress-controller/3.1.x/reference/annotations/ (and here's what it looks like in markdown: https://github.com/Kong/docs.konghq.com/blob/main/app/_src/kubernetes-ingress-controller/reference/annotations.md?plain=1)

> - ```x-kong-tags```: specify the tags to use for each Kong entity generated. The tags can be overridden when doing the conversion. This can only be specified on document level.

Reviewer: ```suggestion
- `x-kong-tags`: Specify the tags to use for each {{site.base_gateway}} entity generated. Tags can be overridden when doing the conversion. This can only be specified at the document level.
```

> - ```x-kong-service-defaults```: the defaults for the [Kong services](/gateway/{{page.release}}/admin-api/#service-object) generated from 'servers' object in the OpenAPI spec. These defaults can also be added to "path" and "operation" objects, in which case a new Service entity will be generated.

Reviewer: ```suggestion
- `x-kong-service-defaults`: The defaults for the [services](/gateway/api/admin-ee/latest/#/Services) generated from the `servers` object in the OpenAPI spec. These defaults can also be added to the `path` and `operation` objects, which will generate a new service entity.
```
Linking to the spec for the admin API instead. We no longer maintain the old markdown pages.

> - ```x-kong-upstream-defaults```: the defaults for [upstreams](https://docs.konghq.com/gateway/latest/admin-api/#upstream-object) generated from the `servers` object in the OpenAPI spec. These defaults can also be added to `path` and `operation` objects, in which case a new Service and Upstream entity will be generated.

Reviewer: ```suggestion
- `x-kong-upstream-defaults`: The defaults for [upstreams](/gateway/api/admin-ee/latest/#/Upstreams) generated from the `servers` object in the OpenAPI spec. These defaults can also be added to the `path` and `operation` objects, which will generate a new service entity.
```

> - ```x-kong-route-defaults```: the defaults for the [Kong routes](/gateway/{{page.release}}/admin-api/#route-object) generated from 'paths' in the OpenAPI spec.

Reviewer: ```suggestion
- `x-kong-route-defaults`: The defaults for the [routes](/gateway/api/admin-ee/latest/#/Routes) generated from `paths` in the OpenAPI spec.
```

> - ```x-kong-name```: the above directive gives the entire spec file its name. This will be used for naming the service and upstream objects in Kong. If not given it will use the 'info.title' field above to name them, or a random UUID if the 'info.title' field is missing. Names are converted into valid identifiers. This directive can also be used on "path" and "operation" objects to name them. Similar to operationId, each x-kong-name must be unique within the spec file.

Reviewer: ```suggestion
- `x-kong-name`: The name for the entire spec file. This is used for naming the service and upstream objects in {{site.base_gateway}}. If not given, it will use the `info.title` field to name these objects, or a random UUID if the `info.title` field is missing. Names are converted into valid identifiers. This directive can also be used on `path` and `operation` objects to name them. Similarly to `operationId`, each `x-kong-name` must be unique within the spec file.
```

> - ```x-kong-name```: the above directive gives the entire spec file its name. This will be used for naming the service and upstream objects in Kong. If not given it will use the 'info.title' field above to name them, or a random UUID if the 'info.title' field is missing. Names are converted into valid identifiers. This directive can also be used on "path" and "operation" objects to name them. Similar to operationId, each x-kong-name must be unique within the spec file.

Reviewer: +1

> - ```x-kong-plugin-{Kong plugin name}```: Directive to add a plugin. The plugin name is derived from the extension name and is a generic mechanism that can add any type of plugin. This plugin is configured on a global level for the OpenAPI spec. As such it will be configured on the Kong Service entity, and hence apply on all paths and operations in this spec. It can be specified again on paths and operations to override the config for that specific subset of the spec, in that case it will be added to the generated Kong Route entity. If new Service entities are generated from "path" or "operation" objects, the plugins will be copied over accordingly (for example by having "servers" objects, or Upstream or Service defaults specified on those levels). A consumer can be referenced by setting the "consumer" field to the consumer name or id. Note: since the plugin name is in the key, only 1 instance of each plugin can be added on each level.

Reviewer: ```suggestion
- `x-kong-plugin-{Kong plugin name}`: Directive to add a plugin. The plugin name is derived from the extension name and is a generic mechanism that can add any type of plugin. This plugin is configured on a global level for the OpenAPI spec. As such, it is configured on the service entity, and applies on all paths and operations in this spec. 

  The plugin name can also be specified on paths and operations to override the config for that specific subset of the spec. In that case, it is added to the generated route entity. If new service entities are generated from `path` or `operation` objects, the plugins are copied over accordingly (for example, by having `servers`  objects, or upstream or service defaults specified on those levels). 
  
  A consumer can be referenced by setting the `consumer` field to the consumer name or ID. 
   
  {:.note}
  > Note: Since the plugin name is in the key, only one instance of each plugin can be added at each level.
```

> - ```x-kong-plugin-{Kong plugin name}```: Directive to add a plugin. The plugin name is derived from the extension name and is a generic mechanism that can add any type of plugin. This plugin is configured on a global level for the OpenAPI spec. As such it will be configured on the Kong Service entity, and hence apply on all paths and operations in this spec. It can be specified again on paths and operations to override the config for that specific subset of the spec, in that case it will be added to the generated Kong Route entity. If new Service entities are generated from "path" or "operation" objects, the plugins will be copied over accordingly (for example by having "servers" objects, or Upstream or Service defaults specified on those levels). A consumer can be referenced by setting the "consumer" field to the consumer name or id. Note: since the plugin name is in the key, only 1 instance of each plugin can be added on each level.

Reviewer: Not in the docs. The rest of the entities here should be lowercase (service, route). We could argue which way is more "correct", but the consistent approach is to use lowecase.

> - ```securitySchemes.[...].x-kong-security-openid-connect```: specifies that the Kong OpenID Connect plugin is to be used to implement this "security scheme object". Any custom configuration can be added as usual for plugins.

Reviewer: ```suggestion
- `securitySchemes.[...].x-kong-security-openid-connect`: Specifies that the [OpenID Connect plugin](/hub/kong-inc/openid-connect) is to be used to implement this `security scheme object`. Any custom configuration can be added as usual for plugins.
```

> - ```components.x-kong```: reusable Kong configuration components. All Kong references must be under this key. Referenceable elements are:

Reviewer: ```suggestion
- `components.x-kong`: Reusable {{site.base_gateway}} configuration components. All `x-kong` references must be under this key. It accepts the following referenceable elements:
```

>   - ```x-kong-service-defaults```
>   - ```x-kong-upstream-defaults```
>   - ```x-kong-route-defaults```
>   - ```x-kong-plugin-[...] plugin configurations```
>   - ```x-kong-security-[...] plugin configurations```

Reviewer: ```suggestion
  - `x-kong-service-defaults`
  - `x-kong-upstream-defaults`
  - `x-kong-route-defaults`
  - `x-kong-plugin-[...] plugin configurations`
  - `x-kong-security-[...] plugin configurations`
```
